### PR TITLE
Add timeout to abort joining of seed nodes, #23502

### PR DESCRIPTION
**Pull request by patriknw**
Refs #23502

@huntc please review

### CODE REVIEW

File: akka-docs/src/main/paradox/scala/cluster-usage.md [hunk 1]
```diff
@@ -111,8 +108,11 @@ The source code of this sample can be found in the
 ## Joining to Seed Nodes
```
Comment (patriknw) [inline akka-docs/src/main/paradox/scala/cluster-usage.md]: @WadeWaldron I tried to incorporate some of your findings in the documentation of the seed nodes. Please review, and let me know if there is anything more or if still unclear.

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala [hunk 2]
```diff
@@ -390,18 +398,37 @@ private[cluster] class ClusterCoreDaemon(publisher: ActorRef) extends Actor with
...
         if (seedNodes.nonEmpty) joinSeedNodes(seedNodes)
         else join(joinWith)
       }
   }: Actor.Receive).orElse(receiveExitingCompleted)
 
+  private def resetJoinSeedNodesDeadline(): Unit = {
+    joinSeedNodesDeadline = ShutdownAfterUnsuccessfulJoinSeedNodes match {
+      case d: FiniteDuration ⇒ Some(Deadline.now + d)
+      case _                 ⇒ None // off
+    }
```
Comment (raboof) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: This seems a bit strange here, is `ShutdownAfterUnsuccessfulJoinSeedNodes` a nullable `Duration`? Perhaps we should make it an `Option[FiniteDuraction]` already on initialization?

File: akka-cluster/src/test/scala/akka/cluster/ShutdownAfterJoinSeedNodesSpec.scala [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    "be aborted after shutdown-after-unsuccessful-join-seed-nodes" taggedAs LongRunningTest in {
+
+      val seedNodes: immutable.IndexedSeq[Address] = Vector(seed1, seed2).map(s ⇒ Cluster(s).selfAddress)
+      shutdown(seed1) // crash so that others will not be able to join
+
+      Cluster(seed2).joinSeedNodes(seedNodes)
+      Cluster(oridinary1).joinSeedNodes(seedNodes)
+
+      Await.result(seed2.whenTerminated, Cluster(seed2).settings.ShutdownAfterUnsuccessfulJoinSeedNodes + 1.second)
+      Await.result(oridinary1.whenTerminated, Cluster(seed2).settings.ShutdownAfterUnsuccessfulJoinSeedNodes + 1.second)
```
Comment (raboof) [inline akka-cluster/src/test/scala/akka/cluster/ShutdownAfterJoinSeedNodesSpec.scala]: Isn't 1 second a bit tight? We'll see I suppose :)

**Review by raboof — APPROVED** [review]
Nice improvement, docs look good.

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala (hunk 2 above)
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: It's `Duration.Undefined` (not finite) when "off", see ClusterSettings.
We have similar in a few other settings here so I followed same pattern.

File: akka-cluster/src/test/scala/akka/cluster/ShutdownAfterJoinSeedNodesSpec.scala (hunk 3 above)
Comment (patriknw) [inline akka-cluster/src/test/scala/akka/cluster/ShutdownAfterJoinSeedNodesSpec.scala]: you are right, no reason to have it tight (the actual shutdown of actor system may take longer)

File: akka-cluster/src/main/resources/reference.conf [hunk 4]
```diff
@@ -15,12 +15,25 @@ akka {
...
     seed-node-timeout = 5s
 
     # If a join request fails it will be retried after this period.
     # Disable join retry by specifying "off".
     retry-unsuccessful-join-after = 10s
+    
+    # The joining of given seed nodes will by default be retried indefinitely until
+    # a successful join. That process can be aborted if unsuccessful by defining this
+    # timeout. When aborted it will run CoordinatedShutdown, which by default will
+    # terminated the ActorSystem. CoordinatedShutdown can also be configured to exit
```
Comment (huntc) [inline akka-cluster/src/main/resources/reference.conf]: Should be just "terminate"

File: akka-docs/src/main/paradox/scala/cluster-usage.md [hunk 5]
```diff
@@ -131,6 +131,12 @@ This can also be defined as Java system properties when starting the JVM using t
 -Dakka.cluster.seed-nodes.1=akka.tcp://ClusterSystem@host2:2552
 ```
 
+Such configuration is typically created dynamically by external tools, see for example:
+
+* [Deploying clustered Akka applications on Kubernetes](http://developer.lightbend.com/guides/k8-akka-cluster/)
+* [ConductR](https://conductr.lightbend.com/docs/2.2.x/AkkaAndPlay#Akka-Clustering)
```
Comment (huntc) [inline akka-docs/src/main/paradox/scala/cluster-usage.md]: Please use the 2.1.x link - 2.1 is the current release.

**Review by huntc — APPROVED** [review]

**Comment by patriknw** [comment]
incorporated feedback, will merge after validation

**Pull request merged**